### PR TITLE
WIP interpreter

### PR DESCRIPTION
**Pull request by carnaval**
Make the interpreter able to handle a lot more julia including function calls, ccalls and a couple more intrinsics. It still makes functions go through inference which is unnecessary but was easier for now.

The way I handle ccall (in particular &ref arguments) is either cute or gore depending on your particular fetishes.

There is a weird thing the frontend does where it emits undefvar nodes for arguments (see Dict's setindex!), it may be a bug.

That being said, it's close to getting a REPL in full interpreted mode, except of course that it cannot handle type_goto (I'm gonna wait for this to go away) and generates compiled little stubs for ccalls (one per signature).

Oh and it's super slow obviously.

@JeffBezanson

### CODE REVIEW

**Comment by carnaval** [comment]
also the code should move around, a big part of it should go into runtime_intrinsics for example

**Comment by StefanKarpinski** [comment]
I wonder how slow running the test suite in the interpreter would be. That would be a good test of it. Some things would run abysmally slow, but then again no compilation overhead! So it might be a wash.

**Comment by quinnj** [comment]
Can we get a reminder of why this would be useful? Helping with #265 perhaps?

**Comment by carnaval** [comment]
the plan is to get the interpreter to be more general (i.e. can run all julia code) and faster (no O(n) variable lookup which Jeff is working on) and then abandon compilation when the function would clearly not benefit from it (i.e. only boxes and `jl_apply_generic`s).

Obviously it would still be a very naive implementation, the state of the art in fast interpreters is close to black art.

File: src/interpreter.c [hunk 1]
```diff
@@ -55,18 +64,132 @@ JL_DLLEXPORT jl_value_t *jl_interpret_toplevel_expr_in(jl_module_t *m,
...
+                result = jl_apply_generic_ex(argv2, 5, 0);
+            } else {
+                jl_value_t *ccall_args = jl_alloc_svec(nargs - 4); // not rooted until next assignment
+                memcpy(jl_svec_data(ccall_args), &argv[4], (nargs - 4)*sizeof(jl_value_t*));
+                argv[4] = ccall_args;
+                result = jl_apply_generic_ex(argv, 5, 0);
+            }
+            jl_interpreter_enabled = was_enabled;
+            return result;
+        } else if (intr_id == 90) { // cglobal implementation
```
Comment (vtjnash) [inline src/interpreter.c]: there should have already been a runtime_intrinsic for this one

Comment (vtjnash) [inline src/interpreter.c]: ah, foey: i never pushed this fix to master: c0dc384 (minus the diff for src/toplevel.c)

Comment (vtjnash) [inline src/interpreter.c]: sigh, nor did it's parent commit (164180c4f2e4fa55e8c153a96b3b8fc8748a5eff)

**Comment by StefanKarpinski** [comment]
> Can we get a reminder of why this would be useful?

Various things:

1. Faster startup.
2. Faster execution for situations where JIT isn't worth it.
3. Static compilation of Julia code. In particular eval and/or generated functions, both of which cannot require JIT or an interpreter. So if you want to get rid of JIT entirely and have full Julia available, you need a full-featured interpreter.
4. Debugging. You may want to set a breakpoint in compiled code and dismount (via on stack replacement) into an interpreter to do the actual debugging since that's much more reliable.

**Comment by Keno** [comment]
> Debugging. You may want to set a breakpoint in compiled code and dismount (via on stack replacement) into an interpreter to do the actual debugging since that's much more reliable.

This is mostly orthogonal. You can get the same by just calling eval on every statement. Of course this might be more efficient.

**Comment by JeffBezanson** [comment]
Cool!

For a bit more background, a small experiment on part of the test suite showed that ~30% of the functions we codegen only call jl_apply_generic, and 50% have no intrinsic calls (but might have specialized call sites). We need to avoid native codegen in these cases.

**Comment by Keno** [comment]
I believe the recent interpreter improvements supersede these changes.

**Pull request closed without merging**